File: app/src/main/java/com/wavesplatform/wallet/v1/request/IssueTransactionRequest.java
package com.wavesplatform.wallet.v1.request;

import android.util.Log;

import com.google.common.primitives.Bytes;
import com.google.common.primitives.Longs;
import com.google.common.primitives.Shorts;
import com.wavesplatform.wallet.v1.crypto.Base58;
import com.wavesplatform.wallet.v1.crypto.CryptoProvider;
import com.wavesplatform.wallet.v1.crypto.Hash;
import com.wavesplatform.wallet.v1.payload.AssetBalance;
import com.wavesplatform.wallet.v1.payload.IssueTransaction;
import com.wavesplatform.wallet.v2.util.AddressUtil;

import org.apache.commons.lang3.ArrayUtils;

public class IssueTransactionRequest {
    public static int MaxDescriptionLength = 1000;
    public static int MinFee = 100000000;
    public static int MaxAssetNameLength = 16;
    public static int MinAssetNameLength = 4;
    public static int MaxDecimals = 8;

    final public String id;
    final public String senderPublicKey;
    final public String name;
    final public String description;
    final public long quantity;
    final public byte decimals;
    final public boolean reissuable;
    final public long fee;
    final public long timestamp;
    public String signature;

    public transient final int txType = 3;
    public transient final byte[] nameBytes;
    public transient final byte[] descriptionBytes;

    public IssueTransactionRequest(String senderPublicKey, String name, String description,
                                   long quantity, byte decimals, boolean reissuable, long timestamp) {
        this.senderPublicKey = senderPublicKey;
        this.nameBytes = name != null ? name.getBytes(org.apache.commons.io.Charsets.UTF_8) : ArrayUtils.EMPTY_BYTE_ARRAY;
        this.name = name;
        this.descriptionBytes = description != null ? description.getBytes(org.apache.commons.io.Charsets.UTF_8) : ArrayUtils.EMPTY_BYTE_ARRAY;
        this.description = description != null ? description : "";
        this.quantity = quantity;
        this.decimals = decimals;
        this.reissuable = reissuable;
        this.fee = MinFee;
        this.timestamp = timestamp;
        this.id = Base58.encode(Hash.fastHash(toSignBytes()));
    }

    private static byte[] arrayWithSize(byte[] b) {
        return Bytes.concat(Shorts.toByteArray((short) b.length), b);
    }

    public byte[] toSignBytes() {
        try {
            byte[] reissuableBytes = reissuable ? new byte[]{1} : new byte[]{0};

            return Bytes.concat(new byte[]{txType},
                    Base58.decode(senderPublicKey),
                    arrayWithSize(nameBytes),
                    arrayWithSize(descriptionBytes),
                    Longs.toByteArray(quantity),
                    new byte[]{decimals},
                    reissuableBytes,
                    Longs.toByteArray(fee),
                    Longs.toByteArray(timestamp));
        } catch (Exception e) {
            Log.e("Wallet", "Couldn't create issue transaction sign", e);
            return new byte[0];
        }
    }

    public void sign(byte[] privateKey) {
        if (signature == null) {
            signature = Base58.encode(CryptoProvider.sign(privateKey, toSignBytes()));
        }
    }
}

File: wavesplatform/src/main/java/com/wavesplatform/sdk/net/model/request/IssueTransactionRequest.kt
/*
 * Created by Eduard Zaydel on 1/4/2019
 * Copyright © 2019 Waves Platform. All rights reserved.
 */

package com.wavesplatform.sdk.net.model.request

import android.util.Log

import com.google.common.primitives.Bytes
import com.google.common.primitives.Longs
import com.google.common.primitives.Shorts
import com.wavesplatform.sdk.crypto.Base58
import com.wavesplatform.sdk.crypto.CryptoProvider
import com.wavesplatform.sdk.crypto.Hash

import org.apache.commons.lang3.ArrayUtils

class IssueTransactionRequest(val senderPublicKey: String, val name: String?, description: String?,
                              val quantity: Long, val decimals: Byte, val reissuable: Boolean, val timestamp: Long) {

    val id: String
    val description: String
    val fee: Long
    var signature: String? = null

    @Transient
    val txType = 3
    @Transient
    val nameBytes: ByteArray
    @Transient
    val descriptionBytes: ByteArray

    init {
        this.nameBytes = name?.toByteArray(org.apache.commons.io.Charsets.UTF_8)
                ?: ArrayUtils.EMPTY_BYTE_ARRAY
        this.descriptionBytes = description?.toByteArray(org.apache.commons.io.Charsets.UTF_8)
                ?: ArrayUtils.EMPTY_BYTE_ARRAY
        this.description = description ?: ""
        this.fee = MinFee.toLong()
        this.id = Base58.encode(Hash.fastHash(toSignBytes()))
    }

    fun toSignBytes(): ByteArray {
        try {
            val reissuableBytes = if (reissuable) byteArrayOf(1) else byteArrayOf(0)

            return Bytes.concat(byteArrayOf(txType.toByte()),
                    Base58.decode(senderPublicKey),
                    arrayWithSize(nameBytes),
                    arrayWithSize(descriptionBytes),
                    Longs.toByteArray(quantity),
                    byteArrayOf(decimals),
                    reissuableBytes,
                    Longs.toByteArray(fee),
                    Longs.toByteArray(timestamp))
        } catch (e: Exception) {
            Log.e("Wallet", "Couldn't create issue transaction sign", e)
            return ByteArray(0)
        }

    }

    fun sign(privateKey: ByteArray) {
        if (signature == null) {
            signature = Base58.encode(CryptoProvider.sign(privateKey, toSignBytes()))
        }
    }

    companion object {
        var MaxDescriptionLength = 1000
        var MinFee = 100000000
        var MaxAssetNameLength = 16
        var MinAssetNameLength = 4
        var MaxDecimals = 8

        private fun arrayWithSize(b: ByteArray): ByteArray {
            return Bytes.concat(Shorts.toByteArray(b.size.toShort()), b)
        }
    }
}
